### PR TITLE
Link condition for tetrahedral mesh

### LINKED ISSUE #30: Additional Test for Collapse
Additional Test for Collapse:
- edges on the boundary, and internal edges connecting the boundary
- test timestamp and revert operations

### LINKED ISSUE #38: add `is_manifold` and `link condition` for edges in 2d and 3d
- check during collapse operation
- Vertex/Face/Edge/Tet Invariants

### PR DESCRIPTION
**Pull request by jiangzhongshi**
Closes #38
Closes #30

- [x] link condition for internals
- [x] link condition for the boundary.
- [x] Clear warnings related to `cmp_l`

### CODE REVIEW

File: src/wmtk/EdgeCollapsingConn.cpp [hunk 1]
```diff
@@ -4,18 +4,140 @@
...
+#include <algorithm>
+#include <cstddef>
+#include <limits>
+#include <type_traits>
 
 bool wmtk::TetMesh::collapse_edge(const Tuple& loc0, std::vector<Tuple>& new_edges)
 {
     if (!collapse_before(loc0)) return false;
 
+    auto link_condition = [&VC = this->m_vertex_connectivity,
```
Comment (teseoch) [inline src/wmtk/EdgeCollapsingConn.cpp:20]: they should be functions inside the class :)

File: src/wmtk/EdgeCollapsingConn.cpp [hunk 2]
```diff
@@ -4,18 +4,140 @@
...
+                           &TC = this->m_tet_connectivity,
+                           this](auto v0, auto v1) -> bool {
+        auto intersects = [](const auto& vec, const auto& val) {
+            for (auto& v : vec)
+                for (auto& a : val)
+                    if (v == a) return true;
+            return false;
+        };
+
+        auto link = [&intersects,
```
Comment (teseoch) [inline src/wmtk/EdgeCollapsingConn.cpp:30]: this too

File: src/wmtk/EdgeCollapsingConn.cpp (hunk 1 above)
Comment (jiangzhongshi) [reply src/wmtk/EdgeCollapsingConn.cpp:20]: It captures a `this`

Comment (jiangzhongshi) [reply src/wmtk/EdgeCollapsingConn.cpp:20]: I would prefer to refactor once it is needed.

File: src/wmtk/EdgeCollapsingConn.cpp (hunk 2 above)
Comment (jiangzhongshi) [reply src/wmtk/EdgeCollapsingConn.cpp:30]: this results in a crazy long function signature.

File: src/wmtk/EdgeCollapsingConn.cpp (hunk 1 above)
Comment (jiangzhongshi) [reply src/wmtk/EdgeCollapsingConn.cpp:20]: Maybe create an issue like "unify 2d and 3d link condition etc."

**Pull request merged**